src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

```diff
@@ -257,6 +258,338 @@ private struct IntPtrAndHandle
+            if (typePtr == IntPtr.Zero)
+                return "null";
+            else if (string.IsNullOrEmpty(marshalerFullName))
+                return "null";
+
+            IntPtrAndHandle tmp = default(IntPtrAndHandle);
+            tmp.ptr = typePtr;
+            var typeHandle = tmp.typeHandle;
+
+            var type = Type.GetTypeFromHandle(typeHandle);
+            var marshalerType = Type.GetType(marshalerFullName) ?? type.Assembly.GetType(marshalerFullName);
+            if (marshalerType == null)
+                return "null";
+
+            var preFilter = GetAndEscapeStringProperty(marshalerType, "FromJavaScriptPreFilter");
+            var postFilter = GetAndEscapeStringProperty(marshalerType, "ToJavaScriptPostFilter");
+
+            var inputPtr = GetMarshalMethodPointer(marshalerType, "FromJavaScript", out Type? fromReturnType, out Type? fromParameterType);
+            var outputPtr = GetMarshalMethodPointer(marshalerType, "ToJavaScript", out Type? toReturnType, out Type? toParameterType);
+
+            if (inputPtr == IntPtr.Zero)
+                throw new WasmInteropException($"{marshalerType.Name} must have a static FromJavaScript method");
+            if (outputPtr == IntPtr.Zero)
+                throw new WasmInteropException($"{marshalerType.Name} must have a static ToJavaScript method");
+
+            if (fromReturnType != type)
+                throw new WasmInteropException($"{marshalerType.Name}.FromJavaScript's return type must be {type.Name} but was {fromReturnType}");
+
+            if (type.IsValueType && (toParameterType != null)) {
+                var typeMatches = toParameterType.GetElementType() == type;
+                if (!typeMatches || !(toParameterType.IsPointer || toParameterType.IsByRef))
+                    throw new WasmInteropException($"{marshalerType.Name}.ToJavaScript's parameter must be 'in {type.Name}' or '{type.Name}*' but was {toParameterType}");
+            } else {
+                if (toParameterType != type)
+                    throw new WasmInteropException($"{marshalerType.Name}.ToJavaScript's parameter must be of type {type.Name} but was {toParameterType}");
+            }
+
+            return ("{\n" + $"\"typePtr\": {typePtr}, \n" +
+                $"\"preFilter\": {preFilter}, \"postFilter\": {postFilter}, \n" +
+                $"\"inputPtr\": {inputPtr}, \"outputPtr\": {outputPtr} \n" +
+                "}");
+        }
+
+        private static MarshalType GetMarshalTypeFromType (Type type) {
+            if (type == null)
+                return MarshalType.VOID;
+            else if (type.IsPointer)
+                return MarshalType.POINTER;
+
+            switch (Type.GetTypeCode(type)) {
+                case TypeCode.Byte:
+                case TypeCode.SByte:
+                case TypeCode.Int16:
+                case TypeCode.UInt16:
+                case TypeCode.Int32:
+                    if (type.IsEnum)
```

Comment:
How does this work for byte-based enums? Can you add tests for enums with different underlying type?

Reply:
Not sure, I will add tests for it. This is an area we didn't cover at all before so it's worth knowing